plugins/grpc-ecosystem/gateway/v2.19.1/separate_pkg_additional_imports.patch

```diff
@@ -195,26 +183,30 @@ index 62766254..033d1c09 100644
  	var imports []descriptor.GoPackage
  	for _, pkgpath := range []string{
  		"context",
-@@ -65,6 +74,7 @@ func New(reg *descriptor.Registry, useRequestContext bool, registerFuncSuffix st
+@@ -65,6 +73,7 @@ func New(reg *descriptor.Registry, useRequestContext bool, registerFuncSuffix st
  		registerFuncSuffix: registerFuncSuffix,
  		allowPatchFeature:  allowPatchFeature,
  		standalone:         standalone,
 +		separatePackage:    separatePackage,
  	}
  }
  
-@@ -90,10 +100,19 @@ func (g *generator) Generate(targets []*descriptor.File) ([]*descriptor.Response
+@@ -90,10 +99,23 @@ func (g *generator) Generate(targets []*descriptor.File) ([]*descriptor.Response
  			grpclog.Errorf("%v: %s", err, code)
  			return nil, err
  		}
 +		goPkg := file.GoPkg
 +		fileNamePrefix := file.GeneratedFilenamePrefix
 +		if g.separatePackage {
 +			goPkg = descriptor.GoPackage{
-+				Path: filepath.Join(file.GoPkg.Path, file.GoPkg.Name),
++				Path: path.Join(file.GoPkg.Path, file.GoPkg.Name),
 +				Name: file.GoPkg.Name,
 +			}
-+			fileNamePrefix = path.Join(file.GeneratedFilenamePrefix, file.GoPkg.Name, filepath.Base(file.GeneratedFilenamePrefix))
```

Comment:
This will unfortunately be a breaking change, but this appears like a bug in the previous implementation. We were including the filename as a directory and then concatenating with the grpcgateway directory:

### Before

```
service1
└── v1
    ├── service
    │   └── service1v1gateway
    │       └── service.pb.gw.go
    ├── service.pb.go
    └── service1v1grpc
        └── service_grpc.pb.go
```

### After

```
service1
└── v1
    ├── service.pb.go
    ├── service1v1gateway
    │   └── service.pb.gw.go
    └── service1v1grpc
        └── service_grpc.pb.go
```

Reply:
Hopefully this won't be too much of an impact to people - the new package layout is definitely what we intended it to be originally and it matches how grpc and connect generate code.

Reply:
I've thought long about this, and as much as I dislike breaking changes, I'd argue this is a bug fix at the expense of a quick import fix for end-users.

It'll also make the grpc-gateway implementation consistent with the grpc and connect plugins. And if we have additional plugins that generate to a separate directory they should all follow a convention, which leaves this grpc-gateway plugin as an outlier.

Lastly, explaining generated SDK paths is sometimes challenging to end-users. Having this non-conventional path is yet another barrier to adoption, at least for grpc-gateway.

cc @bufdev what do you think?

The alternative is to leave the generated imports as-is and remember to fix this in a major version only.

Reply:
@pkwarren reverted the breaking change in https://github.com/bufbuild/plugins/pull/1223/commits/2dba0c1b7fe1bafb999d7556f1ba073284a00a7a

For the gateway, we continue to build the same path as before. I left a comment to remind us of this in case we want to address this in the future.